File: 2023/solutions/day06.py
from input import Input
from math import prod

def parse_input(input: Input) -> list[tuple[int, int]]:
    return list(zip([int(n) for n in input.lines[0].split()[1:]], [int(n) for n in input.lines[1].split()[1:]]))

def ways_to_win(time: int, dst: int) -> int:
    sum = 0
    for i in range(1, time):
        sum += (time - i) * i > dst
    return sum

def solve_first(input: Input) -> object:
    pairs = parse_input(input)
    return prod(ways_to_win(time, dst) for time, dst in pairs)

def solve_second(input: Input) -> object:
    pairs = parse_input(input)
    time = int("".join(str(pair[0]) for pair in pairs))
    dst = int("".join(str(pair[1]) for pair in pairs))
    return ways_to_win(time, dst)

File: 2025/src/day06.fs
module Day06
open Input
open Common

[<Literal>]
let name = "Trash Compactor"

let parseInput<'T> (numMapping: string -> 'T) (input: Input) = 
    let nums = input.lines |> List.take (List.length input.lines - 1) |> List.map numMapping
    let ops = input.lines |> List.last |> Seq.filter ((<>) ' ') |> List.ofSeq
    (nums, ops)

let charsToNum (chars: char list) = 
    chars
    |> List.filter ((<>) ' ')
    |> List.map (fun c -> int64 c - int64 '0') 
    |> List.fold (fun acc n -> acc * 10L + n) 0L

let solve (nums: int64 list list) (ops: char list) = 
    (List.zip nums ops)
    |> List.sumBy (fun (nums, op) ->
        match op with
        | '+' -> List.sum nums
        | '*' -> List.reduce (*) nums
        | _ -> 0L
    )

let solveFirst (input: Input) = 
    let (nums, ops) = parseInput (Parsing.numbers int64) input
    solve (nums |> List.transpose) ops

let solveSecond (input: Input) = 
    let (nums, ops) = parseInput List.ofSeq input
    nums
    |> List.transpose
    |> List.split (List.forall ((=) ' '))
    |> List.map (List.map charsToNum)
    |> fun groups -> solve groups ops
